// @langfuse/otel/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  KNOWN_LLM_INSTRUMENTATION_SCOPE_PREFIXES: () => KNOWN_LLM_INSTRUMENTATION_SCOPE_PREFIXES,
  LangfuseSpanProcessor: () => LangfuseSpanProcessor,
  isDefaultExportSpan: () => isDefaultExportSpan,
  isGenAISpan: () => isGenAISpan,
  isKnownLLMInstrumentor: () => isKnownLLMInstrumentor,
  isLangfuseSpan: () => isLangfuseSpan
});
module.exports = __toCommonJS(index_exports);

// src/span-processor.ts
var import_core3 = require("@langfuse/core");
var import_core4 = require("@opentelemetry/core");
var import_exporter_trace_otlp_http = require("@opentelemetry/exporter-trace-otlp-http");
var import_sdk_trace_base = require("@opentelemetry/sdk-trace-base");

// src/MediaService.ts
var import_core = require("@langfuse/core");
var MediaService = class {
  constructor(params) {
    this.pendingMediaUploads = /* @__PURE__ */ new Set();
    this.apiClient = params.apiClient;
  }
  get logger() {
    return (0, import_core.getGlobalLogger)();
  }
  async flush() {
    await Promise.all(Array.from(this.pendingMediaUploads));
  }
  async process(span) {
    var _a;
    const mediaAttributes = [
      import_core.LangfuseOtelSpanAttributes.OBSERVATION_INPUT,
      import_core.LangfuseOtelSpanAttributes.TRACE_INPUT,
      import_core.LangfuseOtelSpanAttributes.OBSERVATION_OUTPUT,
      import_core.LangfuseOtelSpanAttributes.TRACE_OUTPUT,
      import_core.LangfuseOtelSpanAttributes.OBSERVATION_METADATA,
      import_core.LangfuseOtelSpanAttributes.TRACE_METADATA
    ];
    for (const mediaAttribute of mediaAttributes) {
      const mediaRelevantAttributeKeys = Object.keys(span.attributes).filter(
        (attributeName) => attributeName.startsWith(mediaAttribute)
      );
      for (const key of mediaRelevantAttributeKeys) {
        const value = span.attributes[key];
        if (typeof value !== "string") {
          this.logger.warn(
            `Span attribute ${mediaAttribute} is not a stringified object. Skipping media handling.`
          );
          continue;
        }
        let mediaReplacedValue = value;
        const regex = /data:[^;]+;base64,[A-Za-z0-9+/]+=*/g;
        const foundMedia = [...new Set((_a = value.match(regex)) != null ? _a : [])];
        if (foundMedia.length === 0) continue;
        for (const mediaDataUri of foundMedia) {
          const media = new import_core.LangfuseMedia({
            base64DataUri: mediaDataUri,
            source: "base64_data_uri"
          });
          const langfuseMediaTag = await media.getTag();
          if (!langfuseMediaTag) {
            this.logger.warn(
              "Failed to create Langfuse media tag. Skipping media item."
            );
            continue;
          }
          this.scheduleUpload({
            span,
            media,
            field: mediaAttribute.includes("input") ? "input" : mediaAttribute.includes("output") ? "output" : "metadata"
            // todo: make more robust
          });
          mediaReplacedValue = mediaReplacedValue.replaceAll(
            mediaDataUri,
            langfuseMediaTag
          );
        }
        span.attributes[key] = mediaReplacedValue;
      }
    }
    if (span.instrumentationScope.name === "ai") {
      const aiSDKMediaAttributes = ["ai.prompt.messages", "ai.prompt"];
      for (const mediaAttribute of aiSDKMediaAttributes) {
        const value = span.attributes[mediaAttribute];
        if (!value || typeof value !== "string") {
          continue;
        }
        let mediaReplacedValue = value;
        try {
          const parsed = JSON.parse(value);
          if (Array.isArray(parsed)) {
            for (const message of parsed) {
              if (Array.isArray(message["content"])) {
                const contentParts = message["content"];
                for (const part of contentParts) {
                  if (part["type"] === "file") {
                    let base64Content = null;
                    if (part["data"] != null && part["mediaType"] != null && typeof part["data"] !== "object" && // skip URL instances
                    !String(part["data"]).startsWith("http")) {
                      base64Content = part["data"];
                    }
                    if (part["image"] != null && part["mediaType"] != null && !part["image"].startsWith("http")) {
                      base64Content = part["image"];
                    }
                    if (!base64Content) continue;
                    const media = new import_core.LangfuseMedia({
                      contentType: part["mediaType"],
                      contentBytes: (0, import_core.base64ToBytes)(base64Content),
                      source: "bytes"
                    });
                    const langfuseMediaTag = await media.getTag();
                    if (!langfuseMediaTag) {
                      this.logger.warn(
                        "Failed to create Langfuse media tag. Skipping media item."
                      );
                      continue;
                    }
                    this.scheduleUpload({
                      span,
                      media,
                      field: "input"
                    });
                    mediaReplacedValue = mediaReplacedValue.replaceAll(
                      base64Content,
                      langfuseMediaTag
                    );
                  }
                }
              }
            }
          }
          span.attributes[mediaAttribute] = mediaReplacedValue;
        } catch (err) {
          this.logger.warn(
            `Failed to handle media for AI SDK attribute ${mediaAttribute} for span ${span.spanContext().spanId}`,
            err
          );
        }
      }
    }
  }
  scheduleUpload(params) {
    const { span, field, media } = params;
    const uploadPromise = this.handleUpload({
      media,
      traceId: span.spanContext().traceId,
      observationId: span.spanContext().spanId,
      field
    }).catch((err) => {
      this.logger.error("Media upload failed with error: ", err);
    });
    this.pendingMediaUploads.add(uploadPromise);
    uploadPromise.finally(() => {
      this.pendingMediaUploads.delete(uploadPromise);
    });
  }
  async handleUpload({
    media,
    traceId,
    observationId,
    field
  }) {
    try {
      const contentSha256Hash = await media.getSha256Hash();
      if (!media.contentLength || !media._contentType || !contentSha256Hash || !media._contentBytes) {
        return;
      }
      const { uploadUrl, mediaId } = await this.apiClient.media.getUploadUrl({
        contentLength: media.contentLength,
        traceId,
        observationId,
        field,
        contentType: media._contentType,
        sha256Hash: contentSha256Hash
      });
      if (!uploadUrl) {
        this.logger.debug(
          `Media status: Media with ID ${mediaId} already uploaded. Skipping duplicate upload.`
        );
        return;
      }
      const clientSideMediaId = await media.getId();
      if (clientSideMediaId !== mediaId) {
        this.logger.error(
          `Media integrity error: Media ID mismatch between SDK (${clientSideMediaId}) and Server (${mediaId}). Upload cancelled. Please check media ID generation logic.`
        );
        return;
      }
      this.logger.debug(`Uploading media ${mediaId}...`);
      const startTime = Date.now();
      const uploadResponse = await this.uploadWithBackoff({
        uploadUrl,
        contentBytes: media._contentBytes,
        contentType: media._contentType,
        contentSha256Hash,
        maxRetries: 3,
        baseDelay: 1e3
      });
      if (!uploadResponse) {
        throw Error("Media upload process failed");
      }
      await this.apiClient.media.patch(mediaId, {
        uploadedAt: (/* @__PURE__ */ new Date()).toISOString(),
        uploadHttpStatus: uploadResponse.status,
        uploadHttpError: await uploadResponse.text(),
        uploadTimeMs: Date.now() - startTime
      });
      this.logger.debug(`Media upload status reported for ${mediaId}`);
    } catch (err) {
      this.logger.error(`Error processing media item: ${err}`);
    }
  }
  async uploadWithBackoff(params) {
    const {
      uploadUrl,
      contentType,
      contentSha256Hash,
      contentBytes,
      maxRetries,
      baseDelay
    } = params;
    for (let attempt = 0; attempt <= maxRetries; attempt++) {
      try {
        let parsedHostname;
        try {
          parsedHostname = new URL(uploadUrl).hostname;
        } catch {
          parsedHostname = "";
        }
        const isSelfHostedGcsBucket = parsedHostname === "storage.googleapis.com" || parsedHostname.endsWith(".storage.googleapis.com");
        const headers = isSelfHostedGcsBucket ? { "Content-Type": contentType } : {
          "Content-Type": contentType,
          "x-amz-checksum-sha256": contentSha256Hash,
          "x-ms-blob-type": "BlockBlob"
        };
        const uploadResponse = await fetch(uploadUrl, {
          method: "PUT",
          body: contentBytes,
          headers
        });
        if (attempt < maxRetries && uploadResponse.status !== 200 && uploadResponse.status !== 201) {
          throw new Error(`Upload failed with status ${uploadResponse.status}`);
        }
        return uploadResponse;
      } catch (e) {
        if (attempt === maxRetries) {
          throw e;
        }
        const delay = baseDelay * Math.pow(2, attempt);
        const jitter = Math.random() * 1e3;
        await new Promise((resolve) => setTimeout(resolve, delay + jitter));
      }
    }
  }
};

// src/span-filter.ts
var import_core2 = require("@langfuse/core");
var KNOWN_LLM_INSTRUMENTATION_SCOPE_PREFIXES = [
  import_core2.LANGFUSE_TRACER_NAME,
  "agent_framework",
  "ai",
  "haystack",
  "langsmith",
  "litellm",
  "openinference",
  "opentelemetry.instrumentation.anthropic",
  "strands-agents",
  "vllm"
];
function isLangfuseSpan(span) {
  return span.instrumentationScope.name === import_core2.LANGFUSE_TRACER_NAME;
}
function isGenAISpan(span) {
  return Object.keys(span.attributes).some(
    (attributeKey) => attributeKey.startsWith("gen_ai.")
  );
}
function isKnownLLMInstrumentor(span) {
  const scope = span.instrumentationScope.name;
  return KNOWN_LLM_INSTRUMENTATION_SCOPE_PREFIXES.some(
    (prefix) => scope === prefix || scope.startsWith(`${prefix}.`)
  );
}
function isDefaultExportSpan(span) {
  return isLangfuseSpan(span) || isGenAISpan(span) || isKnownLLMInstrumentor(span);
}

// src/span-processor.ts
var LangfuseSpanProcessor = class {
  /**
   * Creates a new LangfuseSpanProcessor instance.
   *
   * @param params - Configuration parameters for the processor
   *
   * @example
   * ```typescript
   * const processor = new LangfuseSpanProcessor({
   *   publicKey: 'pk_...',
   *   secretKey: 'sk_...',
   *   environment: 'staging',
   *   flushAt: 10,
   *   flushInterval: 2,
   *   mask: ({ data }) => {
   *     // Custom masking logic
   *     return typeof data === 'string'
   *       ? data.replace(/secret_\w+/g, 'secret_***')
   *       : data;
   *   },
   *   shouldExportSpan: ({ otelSpan }) => {
   *     // Full override of default filtering:
   *     // export only spans from specific services
   *     return otelSpan.name.startsWith("my-service");
   *   }
   * });
   * ```
   */
  constructor(params) {
    this.pendingEndedSpans = /* @__PURE__ */ new Set();
    var _a, _b, _c, _d, _e, _f, _g, _h, _i, _j, _k, _l, _m;
    const logger = (0, import_core3.getGlobalLogger)();
    const publicKey = (_a = params == null ? void 0 : params.publicKey) != null ? _a : (0, import_core3.getEnv)("LANGFUSE_PUBLIC_KEY");
    const secretKey = (_b = params == null ? void 0 : params.secretKey) != null ? _b : (0, import_core3.getEnv)("LANGFUSE_SECRET_KEY");
    const baseUrl = (_e = (_d = (_c = params == null ? void 0 : params.baseUrl) != null ? _c : (0, import_core3.getEnv)("LANGFUSE_BASE_URL")) != null ? _d : (0, import_core3.getEnv)("LANGFUSE_BASEURL")) != null ? _e : (
      // legacy v2
      "https://cloud.langfuse.com"
    );
    if (!(params == null ? void 0 : params.exporter) && !publicKey) {
      logger.warn(
        "No exporter configured and no public key provided in constructor or as LANGFUSE_PUBLIC_KEY env var. Span exports will fail."
      );
    }
    if (!(params == null ? void 0 : params.exporter) && !secretKey) {
      logger.warn(
        "No exporter configured and no secret key provided in constructor or as LANGFUSE_SECRET_KEY env var. Span exports will fail."
      );
    }
    const flushAt = (_f = params == null ? void 0 : params.flushAt) != null ? _f : (0, import_core3.getEnv)("LANGFUSE_FLUSH_AT");
    const flushIntervalSeconds = (_g = params == null ? void 0 : params.flushInterval) != null ? _g : (0, import_core3.getEnv)("LANGFUSE_FLUSH_INTERVAL");
    const authHeaderValue = (0, import_core3.base64Encode)(`${publicKey}:${secretKey}`);
    const timeoutSeconds = (_i = params == null ? void 0 : params.timeout) != null ? _i : Number((_h = (0, import_core3.getEnv)("LANGFUSE_TIMEOUT")) != null ? _h : 5);
    const exporter = (_j = params == null ? void 0 : params.exporter) != null ? _j : new import_exporter_trace_otlp_http.OTLPTraceExporter({
      url: `${baseUrl}/api/public/otel/v1/traces`,
      headers: {
        Authorization: `Basic ${authHeaderValue}`,
        "x-langfuse-sdk-name": "javascript",
        "x-langfuse-sdk-version": import_core3.LANGFUSE_SDK_VERSION,
        "x-langfuse-public-key": publicKey != null ? publicKey : "<missing>",
        ...params == null ? void 0 : params.additionalHeaders
      },
      timeoutMillis: timeoutSeconds * 1e3
    });
    this.processor = (params == null ? void 0 : params.exportMode) === "immediate" ? new import_sdk_trace_base.SimpleSpanProcessor(exporter) : new import_sdk_trace_base.BatchSpanProcessor(exporter, {
      maxExportBatchSize: flushAt ? Number(flushAt) : void 0,
      scheduledDelayMillis: flushIntervalSeconds ? Number(flushIntervalSeconds) * 1e3 : void 0
    });
    this.publicKey = publicKey;
    this.baseUrl = baseUrl;
    this.environment = (_k = params == null ? void 0 : params.environment) != null ? _k : (0, import_core3.getEnv)("LANGFUSE_TRACING_ENVIRONMENT");
    this.release = (_l = params == null ? void 0 : params.release) != null ? _l : (0, import_core3.getEnv)("LANGFUSE_RELEASE");
    this.mask = params == null ? void 0 : params.mask;
    this.shouldExportSpan = (_m = params == null ? void 0 : params.shouldExportSpan) != null ? _m : ({ otelSpan }) => isDefaultExportSpan(otelSpan);
    this.apiClient = new import_core3.LangfuseAPIClient({
      baseUrl: this.baseUrl,
      username: this.publicKey,
      password: secretKey,
      xLangfusePublicKey: this.publicKey,
      xLangfuseSdkVersion: import_core3.LANGFUSE_SDK_VERSION,
      xLangfuseSdkName: "javascript",
      environment: "",
      // noop as baseUrl is set
      headers: params == null ? void 0 : params.additionalHeaders
    });
    this.mediaService = new MediaService({ apiClient: this.apiClient });
    logger.debug("Initialized LangfuseSpanProcessor with params:", {
      publicKey,
      baseUrl,
      environment: this.environment,
      release: this.release,
      timeoutSeconds,
      flushAt,
      flushIntervalSeconds
    });
  }
  get logger() {
    return (0, import_core3.getGlobalLogger)();
  }
  /**
   * Called when a span is started. Adds environment, release, and propagated attributes to the span.
   *
   * @param span - The span that was started
   * @param parentContext - The parent context
   *
   * @override
   */
  onStart(span, parentContext) {
    span.setAttributes({
      [import_core3.LangfuseOtelSpanAttributes.ENVIRONMENT]: this.environment,
      [import_core3.LangfuseOtelSpanAttributes.RELEASE]: this.release,
      ...(0, import_core3.getPropagatedAttributesFromContext)(parentContext)
    });
    return this.processor.onStart(span, parentContext);
  }
  /**
   * Called when a span ends. Processes the span for export to Langfuse.
   *
   * This method:
   * 1. Checks if the span should be exported using shouldExportSpan
   *    (custom override or default smart filter)
   * 2. Applies data masking to sensitive attributes
   * 3. Handles media content extraction and upload
   * 4. Logs span details in debug mode
   * 5. Passes the span to the parent processor for export
   *
   * @param span - The span that ended
   *
   * @override
   */
  onEnd(span) {
    const processEndedSpanPromise = this.processEndedSpan(span).catch((err) => {
      this.logger.error(err);
    });
    this.pendingEndedSpans.add(processEndedSpanPromise);
    void processEndedSpanPromise.finally(
      () => this.pendingEndedSpans.delete(processEndedSpanPromise)
    );
  }
  async flush() {
    await Promise.all(Array.from(this.pendingEndedSpans));
    await this.mediaService.flush();
  }
  /**
   * Forces an immediate flush of all pending spans and media uploads.
   *
   * @returns Promise that resolves when all pending operations are complete
   *
   * @override
   */
  async forceFlush() {
    await this.flush();
    return this.processor.forceFlush();
  }
  /**
   * Gracefully shuts down the processor, ensuring all pending operations are completed.
   *
   * @returns Promise that resolves when shutdown is complete
   *
   * @override
   */
  async shutdown() {
    await this.flush();
    return this.processor.shutdown();
  }
  async processEndedSpan(span) {
    var _a, _b;
    try {
      if (this.shouldExportSpan({ otelSpan: span }) === false) {
        this.logger.debug("Dropped span due to shouldExportSpan filter.", {
          spanName: span.name,
          instrumentationScope: span.instrumentationScope.name
        });
        return;
      }
    } catch (err) {
      this.logger.error(
        "shouldExportSpan failed with error. Dropping span.",
        {
          spanName: span.name,
          instrumentationScope: span.instrumentationScope.name
        },
        err
      );
      return;
    }
    await this.applyMaskInPlace(span);
    await this.mediaService.process(span);
    if (this.logger.isLevelEnabled(import_core3.LogLevel.DEBUG)) {
      this.logger.debug(
        `Processed span:
${JSON.stringify(
          {
            name: span.name,
            traceId: span.spanContext().traceId,
            spanId: span.spanContext().spanId,
            parentSpanId: (_b = (_a = span.parentSpanContext) == null ? void 0 : _a.spanId) != null ? _b : null,
            attributes: span.attributes,
            startTime: new Date((0, import_core4.hrTimeToMilliseconds)(span.startTime)),
            endTime: new Date((0, import_core4.hrTimeToMilliseconds)(span.endTime)),
            durationMs: (0, import_core4.hrTimeToMilliseconds)(span.duration),
            kind: span.kind,
            status: span.status,
            resource: span.resource.attributes,
            instrumentationScope: span.instrumentationScope
          },
          null,
          2
        )}`
      );
    }
    this.processor.onEnd(span);
  }
  async applyMaskInPlace(span) {
    const maskCandidates = [
      import_core3.LangfuseOtelSpanAttributes.OBSERVATION_INPUT,
      import_core3.LangfuseOtelSpanAttributes.TRACE_INPUT,
      import_core3.LangfuseOtelSpanAttributes.OBSERVATION_OUTPUT,
      import_core3.LangfuseOtelSpanAttributes.TRACE_OUTPUT,
      import_core3.LangfuseOtelSpanAttributes.OBSERVATION_METADATA,
      import_core3.LangfuseOtelSpanAttributes.TRACE_METADATA
    ];
    for (const maskCandidate of maskCandidates) {
      if (maskCandidate in span.attributes) {
        span.attributes[maskCandidate] = await this.applyMask(
          span.attributes[maskCandidate]
        );
      }
    }
  }
  async applyMask(data) {
    if (!this.mask) return data;
    try {
      return await this.mask({ data });
    } catch (err) {
      this.logger.warn(
        `Applying mask function failed due to error, fully masking property. Error: ${err}`
      );
      return "<fully masked due to failed mask function>";
    }
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  KNOWN_LLM_INSTRUMENTATION_SCOPE_PREFIXES,
  LangfuseSpanProcessor,
  isDefaultExportSpan,
  isGenAISpan,
  isKnownLLMInstrumentor,
  isLangfuseSpan
});
